### PR TITLE
Llu/ln bwd

### PR DESCRIPTION
**Pull request by liqiangxl**
This PR is copy of https://github.com/csarofeen/pytorch/pull/2400
Benchmark results compared with Apex is:
image

### CODE REVIEW

File: test/test_gpu_combined_inner_outer_reduction.cpp
Comment (naoyam) [inline test/test_gpu_combined_inner_outer_reduction.cpp:1]: Let's rename the test file to `test_combined_inner_outer_reduction.cpp`

File: test/test_gpu_combined_inner_outer_reduction.cpp [hunk 1]
```diff
@@ -0,0 +1,957 @@
...
+  float sq_sum = std::inner_product(
+      sub_mean.begin(), sub_mean.end(), sub_mean.begin(), 0.0);
+  float stdev = std::sqrt(sq_sum / nele);
+  return {mean, stdev};
+}
+
+// This case is to test the correctness of the combined inner and outer
+// scheduler used in layer norm backward. It can also be configured to test the
+// performance using different data types.
+TEST_F(NVFuserTest, FusionCombinedSchedulerLayerNormBackward_CUDA) {
```
Comment (naoyam) [inline test/test_gpu_combined_inner_outer_reduction.cpp:45]: We are simplifying test names and the `Fusion` prefix is no longer used.

**Review by naoyam — APPROVED** [review]
LGTM. Just added comments on test naming

**Comment by liqiangxl** [comment]
!build

**Comment by liqiangxl** [comment]
!build

**Comment by liqiangxl** [comment]
!build

**Review by csarofeen — COMMENTED** [review]
@liqiangxl @naoyam why wasn't this implemented as a different scheduler instead of stuffing it into normalization scheduler. You guys know we don't need to schedule every type of fusion in a single scheduler right?

**Comment by naoyam** [comment]
The normalization schedulers are becoming like a collection of related but significantly different schedulers: block-parallel inner normalization, grid-parallel inner normalization (we don't have this yet), block-parallel outer normalization, grid-parallel outer normalization, grid-parallel inner-and-outer normalization (special case for layernorm backward). All of them can be an independent scheduler instead of making them aggregated as the single normalization scheduler. 

I haven't thought about potential implications of the two approaches, but the current aggregated approach seems to make a little more sense for the scheduling performance. If they were all individual schedulers, we would need to call the `canSchedule` of all individual schedulers until the successful scheduler is found, and the `canSchedule` functions of all the normalization scheduler variants would likely to have some common analyses, which would be redundantly executed, like finding persistent buffers and analyzing their sizes.  The common analyses are executed just once in our current design since all of them are just branched out from the single normalization scheduler. The downside is of course the latter definitely makes the scheduler look more unstructured.

I'd say we should keep them as is for now. We would need to rethink about the whole scheduler and segmentation design for more flexibility and composability.

**Comment by csarofeen** [comment]
@liqiangxl could you pull some timing info from some benchmarks to understand how much time is spent in normalization can schedule? I understand the concern @naoyam but it seems like really messy code for the sake of maybe some performance at compilation time.

**Comment by csarofeen** [comment]
PS, you can still have "a scheduler" from the registry perspective that calls into multiple heuristic and scheduling functions.

**Comment by liqiangxl** [comment]
> @liqiangxl could you pull some timing info from some benchmarks to understand how much time is spent in normalization can schedule? I understand the concern @naoyam but it seems like really messy code for the sake of maybe some performance at compilation time.

Sure. The current normalization `canSchedule` will return true if scheduling can be achieved using one of the inner, outer, or combined heuristics. The system is already aware of the specific heuristic to utilize. However, rather than directly accessing the appropriate heuristic, it employs a broader interface, `getPersistentHeuristics`. This interface conducts further analysis and then directs to one of the inner, outer, or combined heuristics. This is also the case for `schedulePersistentKernel`, which serves as a universal interface for the three distinct heuristics. By transitioning from these general interfaces to individual interfaces tailored for each heuristic, we might achieve a cleaner code structure. The change will be like:
image

**Pull request merged**